qiskit/algorithms/optimizers/spsa.py

```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
+
+        return gradient_estimate / resamplings
+
+    def _compute_update(self, loss, x, k, eps):
+        # compute the perturbations
+        if isinstance(self.resamplings, dict):
+            avg = self.resamplings.get(k, 1)
+        else:
+            avg = self.resamplings
+
+        gradient = np.zeros(x.size)
+
+        # accumulate the number of samples
+        deltas = [bernoulli_perturbation(x.size, self.perturbation_dims) for _ in range(avg)]
+
+        gradient = self._point_estimate(loss, x, eps, deltas)
+
+        return gradient
+
+    def _minimize(self, loss, initial_point):
+        # ensure learning rate and perturbation are set
+        # this happens only here because for the calibration the loss function is required
+        if self.learning_rate is None and self.perturbation is None:
+            get_learning_rate, get_perturbation = self.calibrate(loss, initial_point)
+            self.learning_rate = get_learning_rate
+            self.perturbation = get_perturbation
+
+        if self.learning_rate is None or self.perturbation is None:
+            raise ValueError('If one of learning rate or perturbation is set, both must be set.')
+
+        # get iterator
+        eta = self.learning_rate()
+        eps = self.perturbation()
+
+        # prepare some initials
+        x = np.asarray(initial_point)
+
+        self._nfev = 0
+
+        # if blocking is enabled we need to keep track of the function values
+        if self.blocking:
+            fx = loss(x)
+
+            self._nfev += 1
+            if self.allowed_increase is None:
+                self.allowed_increase = 2 * self.estimate_stddev(loss, x)
+
+        logger.info('=' * 30)
+        logger.info('Starting SPSA optimization')
+        start = time()
+
+        # keep track of the last few steps to return their average
+        last_steps = deque([x])
+
+        for k in range(1, self.maxiter + 1):
+            iteration_start = time()
+            # compute update
+            update = self._compute_update(loss, x, k, next(eps))
+
+            # trust region
+            if self.trust_region:
+                norm = np.linalg.norm(update)
+                if norm > 1:  # stop from dividing by 0
+                    update = update / norm
+
+            # compute next parameter value
+            update = update * next(eta)
+            x_next = x - update
+
+            # blocking
+            if self.blocking:
+                fx_next = loss(x_next)
+
+                self._nfev += 1
+                if fx + self.allowed_increase <= fx_next:  # accept only if loss improved
+                    if self.callback is not None:
+                        self.callback(self._nfev,  # number of function evals
+                                      x_next,  # next parameters
+                                      fx_next,  # loss at next parameters
+                                      np.linalg.norm(update),  # size of the update step
```

Comment:
The norm here depends on the dim of the vector.  Since the steps all the same, can one not just take the inf norm as the norm used?

Reply:
Yes you could for standard SPSA. However that's not true anymore for 2-SPSA and QN-SPSA (see links in the description) since the gradient is multiplied with the inverse Hessian. Since we want to add these algos we already used the L2 norm here.

Reply:
Ok.  I have no experience with the second-order method.